exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/transactions/TransactionApi.kt

```diff
@@ -76,12 +76,12 @@ interface TransactionManager {
         fun closeAndUnregister(database: Database) {
             val manager = registeredDatabases[database]
             manager?.let {
-                registeredDatabases.remove(database)
                 databases.remove(database)
                 currentDefaultDatabase.compareAndSet(database, null)
                 if (currentThreadManager.get() == it) {
```

Comment:
@Tapac I tried your fix on master but it didn't fix the issue.

The issue is shown in the test `H2Tests`. When you call `closeAndRegister` from a different `Thread` the `ThreadLocal` is empty, and will thus try to initialize itself. Resulting in `registeredDatabases[database]` begin called which throw `NoSuchElementException` since we remove the `database` from the `registeredDatabases` map on line `79`.

To fix this we could move the removal of `database` to after the `if` check but that would redundantly initialize the `threadLocal`. There are a couple of solutions for that, but they probably introduce similar overhead as just running the initializer and immediately removing it.